Post-SoD refactor: remove pflag, runahead pool

hjoliver opened the pull request:
This is a ~~small~~ SOD follow-up change (no specific Issue, but has been discussed variously)

E.g., partially addresses #3874:
> tasks are spawned "on demand" when ready to run according to the graph. This brings runhead limiting into line with xtriggers, task hold, and queues: it's just another way to temporarily hold back tasks that are otherwise ready to run.
...  we should get rid of the hidden runahead pool and unify runahead limiting with the other limiting mechanisms:

- Make runahead limiting consistent with other forms of limiting (see above)
   - Remove the "runahead task pool". Runahead tasks are now held in the main pool, with an `is_runahead` task indicator (c.f. `is_held`, `is_queued`) 
   - Runahead-limited tasks can now be exposed in the UI, like queued and held tasks, to show the user "this task is ready to run but is being held back by the runahead limit (done for `cylc tui` on this branch).
 
 - A new "hidden pool" is used to hide partially-satisfied tasks (in order to get spawn-on-demand from spawn-on-outputs) 
   
- Remove an annoying vestige of the old SOS dependency matching process
  -  We used to set `task_events_mgr.pflag` ("processing required flag") whenever a task event occurred that required a new round of dependency matching, then if the flag was set we would iterate through the task pool and do dependency matching and other things.
  - Generally tidy up and make clearer the aforementioned "other things" which are now done directly without relying on the old flag

- [UPDATE] restores correct force-triggering behaviour in the presence of queues (broken since spawn-on-demand) implementation: triggering an unqueued tasks queues it; triggering a queued task causes it to run.

`cylc tui` showing queue- and runahead-limited tasks:
![image](https://user-images.githubusercontent.com/2362137/112223659-1bd3a800-8c8f-11eb-9123-eb489a20589d.png)

<!-- The following requirements must be satisfied (with "[x]"). -->
<!-- Mark the PR as a Draft if all requirements are not yet satisfied. -->

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Already covered by existing tests.
- [x] Appropriate change log entry included. 
      - (mention visible runahead tasks) 
- [ ] (master branch) I have opened a documentation PR at cylc/cylc-doc/pull/XXXX.
      - (mention visible runahead tasks, somewhere...) 
- [x] No dependency changes.

hjoliver commented:
(I've been working on this on another branch, not pushed to the PR yet; soon...)

hjoliver commented:
(Tests passing, after some rebase nightmares; I'll update the description and un-Draft this tomorrow/Wednesday)

hjoliver commented:
(Bumped to beta-1 milestone)

hjoliver commented:
(rebased)

hjoliver commented:
Temporarily converted to Draft, pending fix of post-rebase test fails.

hjoliver commented:
(Tests failing after a moderately nasty conflict resolution :angry: )

hjoliver commented:
Tests passing again :tada:

MetRonnie commented on cylc/flow/broadcast_mgr.py:
```diff
@@ -59,12 +59,23 @@ def __init__(self, workflow_db_mgr, data_store_mgr):
             ext_trigger = ext_trigger_queue.get_nowait()
             self.ext_triggers.setdefault(ext_trigger, 0)
             self.ext_triggers[ext_trigger] += 1
+        return set(
+            [
+                itask for itask in itasks
+                if self._match_ext_trigger(itask)
+            ]
+        )
```
```suggestion
        return set(
            itask for itask in itasks
            if self._match_ext_trigger(itask)
        )
```

hjoliver replied:
(Fixed by @oliver-sanders ' PR now anyway).

MetRonnie commented on cylc/flow/xtrigger_mgr.py:
```diff
@@ -357,16 +356,28 @@ def callback(self, ctx: SubFuncContext):
-        for itask in itasks:
-            if itask.state.xtriggers:
-                self.satisfy_xtriggers(itask)
+        satisfied = set(
+            [
+                itask for itask in itasks
+                if itask.state.xtriggers_all_satisfied()
+            ]
+        )
```
```suggestion
        satisfied = set(
            itask for itask in itasks
            if itask.state.xtriggers_all_satisfied()
        )
```

oliver-sanders replied:
Or better still use a native set comprehension `satisfied = { itask for itask ...`

hjoliver replied:
Doh, welcome to Python 3 :rofl:

MetRonnie reviewed (commented):
Not a full review, just had a read through as it will likely affect my work on #3743

dpmatthews reviewed (approved):
(just approving the description - not tested or reviewed!)

dwsutherland commented on cylc/flow/task_pool.py:
```diff
@@ -1268,24 +1311,31 @@ def force_trigger_tasks(self, items, reflow=False):
-                LOG.critical('setting %s ready to run', itask)
-                itask.state.set_prerequisites_all_satisfied()
-                self.data_store_mgr.delta_task_prerequisite(itask)
-                self.data_store_mgr.delta_task_outputs(itask)
```
I guess the change and outputs/prereqs on force trigger doesn't happen in SOD?

I notice these are still called elsewhere..

hjoliver replied:
Hmmm, forgotten the details, I'd better check this change.

hjoliver replied:
(I remember now: force trigger means "run even though prerequisites are not satisfied" - so why pretend they were satisfied? ... hence, there is no data-store update to do here).

dwsutherland reviewed (approved):
Looks good 👍 

Will test after the rebase, so will review my approval after a play.

hjoliver commented on cylc/flow/scheduler.py:
```diff
@@ -1431,12 +1405,66 @@ async def main_loop(self):
-                self.process_task_pool()
+            # Get tasks with newly-satisfied external triggers.
+            check_if_ready = self.broadcast_mgr.check_ext_triggers(
+                [x for x in self.pool.get_tasks()
```
Yeah, we have discussed reformulating the task pool to a dict (it's mentioned in an issue somewhere) to avoid some iteration.  Beyond the scope of this PR though!

hjoliver replied:
(Reformulating the task pool is definitely on the table, as a possible future enhancement).

hjoliver commented:
(feedback addressed, except for one thing to follow up on, and conflicts).

hjoliver commented:
(There are no conflicts on this branch right now so long as you don't select "Rebase and merge".)

MetRonnie commented on cylc/flow/task_pool.py:
```diff
@@ -574,116 +610,121 @@ def remove(self, itask, reason=""):
+            itask.state.reset(is_queued=True)
+            # TODO Reset manual trigger flag. One manual trigger queues and
+            # unqueued task, another one triggers a queued task.
+            self.data_store_mgr.delta_task_state(itask)  # TODO needed?
```
This was marked as resolved, but looks like it might not be?

hjoliver replied:
Another conflict regression, I think.  Good spotting.

hjoliver replied:
Fixed.

MetRonnie commented on cylc/flow/task_pool.py:
```diff
@@ -585,116 +622,121 @@ def remove(self, itask, reason=""):
+                to_queue.append(itask)
 
-    def _release_tasks(self):
+        self.queue_tasks(to_queue)
```
Why does the "private" `self._queue_tasks()` method call the "public" `self.queue_tasks()` and not the other way round?

hjoliver replied:
`_queue_tasks()` is only called by `reload_taskdefs()`. It figures out which tasks need to be queued, then calls the public `queue_tasks()` to do it.  I'll inline the private one into `reload_taskdefs()`, there's no good reason to have it as a separate method.

hjoliver replied:
(`_queue_tasks()` used to be called in the main loop, as well as at reload, to iterate the pool and find tasks to be queued. But now have event-driven queuing of tasks so that's not neded).

MetRonnie commented on cylc/flow/task_pool.py:
```diff
@@ -1285,24 +1329,30 @@ def force_trigger_tasks(self, items, reflow=False):
+            itask = self.get_task_main(name, point, flow_label)
             if itask is not None:
+                # Trigger existing task proxy
+                LOG.critical('setting %s ready to run', itask)
```
Why is this `critical`? Leftover debugging?

hjoliver replied:
No, this has been (wrongly) `critical` for ages. It's not really a change on this branch, just part of a cut-and-paste move of several lines to another location.

However, I actually changed it to `info` myself in f01c4233c0 from #4218 so I'm a bit confused as to why it's shown up as `critical` again here. Anyhow, I'll (re)fix it.

MetRonnie commented on cylc/flow/task_pool.py:
```diff
@@ -1332,7 +1382,14 @@ def sim_time_check(self, message_queue):
 
-    def set_expired_task(self, itask, now):
+    def set_expired_tasks(self):
+        res = False
+        for itask in self.get_tasks():
+            if self._set_expired_task(itask):
+                res = True
+        return res
```
```suggestion
        for itask in self.get_tasks():
            if self._set_expired_task(itask):
                return True
        return False
```

oliver-sanders replied:
Ooh, damn, I didn't spot that one.

hjoliver replied:
Yikes, I apologize to every coder in the world for this abomination :face_with_head_bandage:

MetRonnie replied:
Wait, my bad, didn't think about the side effects of `self._set_expired_task()`

MetRonnie reviewed (commented):
9/29 files viewed. Seems to be a couple of unresolved point from earlier

oliver-sanders reviewed (commented):
There are some quick changes that will reduce the impact of task iteration, have dumped them in a PR:

https://github.com/hjoliver/cylc-flow/pull/16#pullrequestreview-670335428

hjoliver commented on cylc/flow/scheduler.py:
```diff
@@ -1222,8 +1209,8 @@ def run_event_handlers(self, event, reason):
-    def process_task_pool(self):
-        """Queue and release tasks, and submit task jobs.
+    def release_queued_tasks(self):
+        """Release queued tasks, and submit task jobs.
```
Could use "pop" I guess.

hjoliver replied:
(But, out of scope for this PR)

hjoliver commented on cylc/flow/task_pool.py:
```diff
@@ -712,37 +754,17 @@ def _release_tasks(self):
             itask.waiting_on_job_prep = True
             self.data_store_mgr.delta_task_state(itask)
             self.data_store_mgr.delta_task_queued(itask)
-        if released:
-            LOG.debug(
-                "Queue released:\n"
-                + '\n'.join(f"* {r.identity}" for r in released)
-            )
```
This involves less iteration, since we're already iterating over the list for the preceding lines. Do you have a strong preference either for debug vs info on this? (It's debatable IMO).

hjoliver commented on cylc/flow/task_pool.py:
```diff
@@ -928,13 +946,17 @@ def warn_stop_orphans(self):
+        only unhandled failed tasks and un-released runahead tasks.
+
         """
+        self.release_runahead_tasks()
```
Sorry, I don't follow?

This is here in case there are runahead-limited tasks that could be released (because they are now below the runahead limit) at the time of stall check, in which case, we're not stalled.  Unfortunately I don't recall why I thought that was necessary at this point. ...  I'll go back and see if I can see why.

hjoliver commented on cylc/flow/scheduler.py:
```diff
@@ -672,8 +666,7 @@ def _load_pool_from_point(self):
             parent_points = tdef.get_parent_points(point)
             if not parent_points or all(
                     x < self.config.start_point for x in parent_points):
-                initial_tasks.append(TaskID.get(tdef.name, point))
```
(Not used anywhere).

hjoliver commented on cylc/flow/scheduler.py:
```diff
@@ -1441,12 +1413,66 @@ async def main_loop(self):
+            check_if_ready = self.broadcast_mgr.check_ext_triggers(
+                [x for x in self.pool.get_tasks()
+                    if x.state(TASK_STATUS_WAITING)
+                    and not x.state.is_queued
```
No reason that I can think of. There shouldn't be state changes between each iteration. I've extensively refactored this bit along the lines you've suggested. See what you think.

hjoliver commented on cylc/flow/task_proxy.py:
```diff
@@ -63,9 +63,6 @@ class TaskProxy:
         .late_time:
             Time in seconds since epoch, beyond which the task is considered
             late if it is never active.
-        .manual_trigger:
-            Has this task received a manual trigger command? This flag is reset
-            on trigger.
```
Possibly to do with the distinction between force triggering queued and unqueued tasks (broken on current master). I've now fixed that on this branch (without restoring this flag).

hjoliver commented:
Apologies reviewers!  In addition to my several embarrassing crimes again good iterative form (hopefully I would have found the worst of those myself if I had reviewed my own code properly!) ... there were some `TODO`s indicating unfinished work on this branch, which I had neglected to come back to.  I have now done that, plus addressed all review comments, plus this:

- restored correct force-triggering behaviour in the presence of queues (broken since spawn-on-demand) implementation: triggering an unqueued tasks queues it; triggering a queued task causes it to run.

Hopefully good to go now :crossed_fingers:

hjoliver commented on cylc/flow/task_pool.py:
```diff
@@ -303,40 +302,32 @@ def release_runahead_tasks(self):
-                ):
-                    release.append(itask)
-        for itask in release:
+        for itask in (
+            itask
+            for cycle in self.main_pool.values()
+            for itask in cycle.values()
+            if itask.state.is_runahead
+            if itask.state(
+                TASK_STATUS_FAILED,
+                TASK_STATUS_SUCCEEDED,
+                TASK_STATUS_EXPIRED
+            )
+            or itask.is_manual_submit
+        ):
```
I used to agree, although I've pretty much got my eye in these constructs now.  In any case, you can blame @oliver-sanders ' side-PR for this one!  Oliver, why exactly do you prefer this kind of thing over nested loops with conditional blocks in them? I doubt that comprehensions are any more efficient unless they work on basic types.

hjoliver replied:
(However, the variable `cycle` in this one is badly named, I'll fix that).

oliver-sanders replied:
> I used to agree

Ditto, however, once I got used to comprehensions I changed my mind, I think they are more readable and the filtering is much clearer (especially than the `if <cond>: continue` pattern (negative logic etc).

hjoliver commented:
Three approvals should do it!
